### DISCUSSION TITLE
Set `optional` / `nullable` on Picked Schema

### DISCUSSION THREAD

**Discussion by AaronFreemantle** [colinhacks/zod#2156] [Q&A] [answered]
Currently trying to DRY up a lot of our types used to describe our API responses, and have been looking for a way to set the base types as optional/nullable/nullish on a case by case basis.

We start with a base object
```ts
export const BaseUser = z.object({
    firstName: z.string(),
    lastName: z.string(),
    email: z.string(),
});
```

Then pick from these to match what the API will return.
```ts
const UserAPI = BaseUser.pick({
    firstName: true,
    lastName: true,
});
```

Which works great, but say for example we have another API that returns something like:
```ts
const OptionalUserAPI = z.object({
    firstName: z.string.optional(),
    lastName: z.string.nullable(),
    email: z.string.nullish(),
});
```

Is there a way to derive this from the base type and add just the `optional` / `nullable` / `nullish` later?

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
const OptionalUserAPI = z.object( {
    firstName: z.string().optional(),
    lastName: z.string().nullable(),
    email: z.string().nullish(),
} )

const UserAPI = OptionalUserAPI.pick( {
    firstName: true,
    lastName: true,
} )
type UserAPI = z.infer<typeof UserAPI>
// type UserAPI = {
//     lastName: string | null
//     firstName?: string | undefined
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by AaronFreemantle** [reply]
This would work but defeats the purpose. The idea is to not need to redefine the base attribute types for each API, but be able to modify them in place. Sorry I don't think I was super clear in my question.

In this example like to reference the `BaseUser` when writing the definition of the `OptionalUserAPI`. Something like: 

```ts
const OptionalUserAPI = BaseUser.pick({
    firstName: z.optional(),
    lastName: z.nullable(),
})
```
Which I know doesn't exist, but is there a way to use `extend` / `transform` or similar to have the same effect?

```ts
const OptionalUserAPI = BaseUser.pick({
    firstName: true,
    lastName: true,
}).extend({
    firstName: z.string.optional(), // Would work but we have duplication
    lastName: current.nullable(), // Also doesn't exist but something to this effect
})
```

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
const BaseUser = z.object( {
    firstName: z.string(),
    lastName: z.string(),
    email: z.string(),
} )

const OptionalUserAPI = z.object( {
    firstName: BaseUser.shape.firstName.optional(),
    lastName: BaseUser.shape.lastName.nullable(),
    email: BaseUser.shape.email.nullish(),
} )

type OptionalUserAPI = z.infer<typeof OptionalUserAPI>
// type OptionalUserAPI = {
//     lastName: string | null
//     firstName?: string | undefined
//     email?: string | null | undefined
// }
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by timjam** [comment]
I am facing a similar problem. I have defined some base schemas and I am trying to reuse them as much as possible, so that I don't need to redefine all the keys and their values again and again. I can change some properties to optional with `.partial()`, but is there a way to similarly set properties to nullable or nullish?

```typescript

const BaseUser = z.object({
  firstName: z.string(),
  lastName: z.string()
})

type BaseUser = z.infer<typeof BaseUser>
// type BaseUser = {
//     firstName: string
//     lastName: string
// }

const PartialUser = z.object({})
  .merge(BaseUser.partial({ lastName: true }))

type PartialUser = z.infer<typeof PartialUser>
// type PartialUser = {
//    firstName: string
//    lastName?: string | undefined
// }

//
// Now this is what I would like to achieve
//

const PartialButForNullUser = z.object({})
  .merge(BaseUser.nullPartial({ lastName: true }))

type PartialButForNullUser = z.infer<typeof PartialButForNullUser>
// type PartialButForNullUser = {
//    firstName: string
//    lastName: string | null
// }
```

**Reply by JacobWeisenburger** [reply]
Not that I am aware of.
